### PR TITLE
AF: T_max=75 cosine schedule with EMA+vol-10x champion config

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

T_max=75 is an untested gap in the AirfRANS cosine schedule sweep. T_max=50 is the current champion (surface_mse=0.000296, vol_mse=0.002039). T_max=100 was tested early — but that was pre-EMA and pre-vol-10x, so it is not a fair comparison.

A longer cosine cycle at T_max=75 could allow the model to settle deeper into loss basins before each restart, while still maintaining enough restart frequency to escape shallow local minima. This is especially relevant for volume loss improvement, where the model needs sustained pressure-free learning periods. The gap between our vol_mse (0.002039) and the external target (0.0017) is 1.20x — volume is the remaining frontier.

## Current Baseline (target to beat)

From PR #3135 (nezuko/af-ema-vol-weight-10x):

| Metric | Value | Epoch |
|---|---|---|
| surface_mse | **0.000296** | 704 |
| vol_mse | **0.002039** | 743 |

- W&B run: `sh2zyfwr`
- Config: 2L/256d/4H, AdamW lr=6e-4, T_max=50, gc=1.0, WD=1e-2, EMA=0.999, vol-weight=10x, Fourier

External targets: surface < 0.0043 (already crushed), vol < 0.0017 (1.20x gap remains).

## Known Bug Fix

There is a `primary_metric_key` shadowing bug in `train.py` where a local variable (line ~1646-1648) shadows the function (line ~1428), causing an `UnboundLocalError`. **Fix by renaming the local variable to `best_tracking_metric_key`** if you encounter this error.

## Trial 1 — T_max=75, vol-weight=10x, EMA=0.999

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 75 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --ema-decay 0.999 \
  --vol-loss-weight 10.0 \
  --wandb_group af-tmax-75
```

## Trial 2 — T_max=75, vol-weight=12x, EMA=0.999

Run this **only if Trial 1 is stable and converging**. Slightly higher vol-weight may push vol_mse lower given the longer settling period between restarts.

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 75 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --ema-decay 0.999 \
  --vol-loss-weight 12.0 \
  --wandb_group af-tmax-75
```

## What to Report

Report both `val_primary/surface_mse` and `vol_mse` at the best validation epoch for each trial. Include W&B run IDs.

**Target:** surface_mse < 0.000296 AND/OR vol_mse < 0.002039

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT hinata:

## Results

### Trial 1 — T_max=75, vol-weight=10x, EMA=0.999

| Metric | Baseline (PR #3135) | Trial 1 | Delta | Epoch |
|---|---|---|---|---|
| **val_primary/surface_mse** | **0.000296** | **0.000330** | +11.5% ❌ | 737 |
| **full_val/vol_mse** | **0.002039** | **0.002752** | +35.0% ❌ | 737 |
| best vol_mse (any epoch) | 0.002039 | 0.002559 | +25.5% ❌ | 755 |
| test_primary/surface_mse | — | 0.000914 | — | 737 |
| test vol_mse | — | 0.003374 | — | 737 |

- **Total epochs:** 766 (360-min timeout)
- **W&B run:** [`cc8uhord`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/cc8uhord) (hinata/af-tmax75-vol10x)
- **Command:**
  ```bash
  cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
    --dataset airfrans --airfrans-task full \
    --optimizer adamw --lr 6e-4 \
    --cosine-t-max 75 \
    --grad-clip 1.0 --weight-decay 1e-2 \
    --enable-fourier \
    --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
    --epochs 999 --ema-decay 0.999 \
    --volume-loss-weight 10.0 \
    --agent hinata --wandb-name "hinata/af-tmax75-vol10x" \
    --wandb-group af-tmax-75
  ```

External targets: surface < 0.0043 (✅ crushed at 0.000330), vol < 0.0017 (❌ gap remains at 1.62x).

### Trial 2 — T_max=75, vol-weight=12x, EMA=0.999

| Metric | Baseline (PR #3135) | Trial 2 | Delta | Epoch |
|---|---|---|---|---|
| **val_primary/surface_mse** | **0.000296** | **0.001187** | +301% ❌ | 400 |
| **full_val/vol_mse** | **0.002039** | **0.004253** | +109% ❌ | 409 |
| test_primary/surface_mse | — | 0.000940 | — | 400 |
| test vol_mse | — | 0.005840 | — | 400 |

- **Total epochs:** 770 (360-min timeout), but **diverged catastrophically at epoch ~415**
- **Divergence:** surface_mse went from 0.002 → 0.137 in 3 epochs (ep414→417), never recovered. Model settled at ~0.88 surface_mse for remaining 350 epochs. EMA preserved best checkpoint at ep400.
- **W&B run:** [`hx8zz843`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/hx8zz843) (hinata/af-tmax75-vol12x)
- **Command:**
  ```bash
  cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py \
    --dataset airfrans --airfrans-task full \
    --optimizer adamw --lr 6e-4 \
    --cosine-t-max 75 \
    --grad-clip 1.0 --weight-decay 1e-2 \
    --enable-fourier \
    --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
    --epochs 999 --ema-decay 0.999 \
    --volume-loss-weight 12.0 \
    --agent hinata --wandb-name "hinata/af-tmax75-vol12x" \
    --wandb-group af-tmax-75
  ```

### What happened

**Neither trial beat the baseline.** T_max=75 is worse than T_max=50 for AirfRANS under this configuration.

- **Trial 1 (vol-10x)** converged stably through all 766 epochs and reached surface_mse=0.000330 — respectable but 11.5% behind the T_max=50 baseline (0.000296). Volume was 25-35% worse. The longer cosine cycle did not help the model find deeper basins; instead, the less frequent restarts appear to reduce the effective exploration that T_max=50's more frequent cycling provides.

- **Trial 2 (vol-12x)** diverged at epoch ~415. The higher volume loss weight (12x vs 10x) made the optimization landscape more unstable at cosine LR peaks. Before divergence, it was also behind Trial 1 on surface (0.001187 vs 0.000690 at ep400), suggesting vol-weight=12x hurts surface optimization even when stable.

- **Key insight:** The T_max=50 champion already sits at a near-optimal cosine schedule for AirfRANS. T_max=75's longer cycles provide less exploration (fewer restarts per training budget) without the compensating benefit of deeper settling. The T_max=50 sweet spot appears to be dataset-specific — AirfRANS benefits from more frequent restarts, unlike TandemFoil where T_max=150 is optimal.

### Suggested follow-ups

1. **T_max=60** — the gap between T_max=50 (champion) and T_max=75 is large; T_max=60 could probe whether there's any benefit to a slightly longer cycle
2. **T_max=75 + gc=0.5** — tighter gradient clipping might prevent the divergence seen in Trial 2 and could help Trial 1 converge deeper
3. **vol-weight=10x + T_max=50 + gc=0.5** — combining the champion config with softer clipping (a la DrivAerML's gc=0.5 success) to push volume lower

**Note:** The PR commands used `--vol-loss-weight` which is not a valid CLI flag — the correct flag is `--volume-loss-weight`. I used the correct flag for both trials.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3241 because ADVISOR: Closing — T_max=75 surface=0.000330 (+11.5%), vol=0.002559 (+25.5%) vs baseline. Longer cosine cycle is strictly worse than T_max=50 champion. Hinata reassigned..
